packages/block-library/src/footnotes/format.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * External dependencies
+ */
+import { v4 as createId } from 'uuid';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { formatListNumbered as icon } from '@wordpress/icons';
+import { insertObject } from '@wordpress/rich-text';
+import {
+	RichTextToolbarButton,
+	store as blockEditorStore,
+} from '@wordpress/block-editor';
+import { useSelect, useDispatch, useRegistry } from '@wordpress/data';
+import { createBlock } from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import { name } from './block.json';
+
+export const formatName = 'core/footnote';
+export const format = {
+	title: __( 'Footnote' ),
+	tagName: 'a',
+	className: 'fn',
+	contentEditable: false,
+	edit: function Edit( { value, onChange, isObjectActive } ) {
+		const registry = useRegistry();
+		const {
+			getSelectedBlockClientId,
+			getBlockRootClientId,
+			getBlockName,
+			getBlocks,
+		} = useSelect( blockEditorStore );
+		const { selectionChange, insertBlock } =
+			useDispatch( blockEditorStore );
+		function onClick() {
+			registry.batch( () => {
+				const id = createId();
+				const newValue = insertObject( value, {
+					type: formatName,
+					attributes: {
+						href: '#' + id,
+						id: `${ id }-link`,
+						'data-fn': id,
+					},
+					innerHTML: '*',
+				} );
+				newValue.start = newValue.end - 1;
+
+				onChange( newValue );
+
+				const flattenBlocks = ( blocks ) =>
+					blocks.reduce(
+						( acc, block ) => [
+							...acc,
+							block,
+							...flattenBlocks( block.innerBlocks ),
+						],
+						[]
+					);
```

Comment:
This has better semantics than `reduce`:

```suggestion
				function flattenBlocks( blocks ) {
					return blocks.flatMap( ( block ) => [
						block,
						...flattenBlocks( block.innerBlocks ),
					] );
				}
```

Love me a good `flatMap`.